File: Python/BinarySearchTree.py
class TreeNode:

    def __init__(self, val, left = None, right = None, parent = None):
        self.val        = val
        self.leftChild  = left
        self.rightChild = right
        self.parent     = parent

    def hasLeftChild(self):
        return self.leftChild

    def hasRightChild(self):
        return self.rightChild

    def isRoot(self):
        return not self.parent

    def isLeaf(self):
        return not(self.leftChild or self.rightChild)


class BinarySearchTree:

    def __init__(self):
        self.root = None


    def insert(self, val):
        if self.root:
            self._insert(val, self.root)
        else:
            self.root = TreeNode(val)

    def _insert(self, val, currentNode):
        if val < currentNode.val:
            if currentNode.hasLeftChild():
                self._insert(val, currentNode.leftChild)
            else:
                currentNode.leftChild = TreeNode(val, parent = currentNode)
        else:
            if currentNode.hasRightChild():
                self._insert(val, currentNode.rightChild)
            else:
                currentNode.rightChild = TreeNode(val, parent = currentNode)



    # if BST contains node return it

    def contains(self, val):
        if self.root:
            return self._contains(val, self.root)
        else:
            return None

    def _contains(self, val, currentNode):
        if not currentNode:
            return None
        elif val == currentNode.val:
            return currentNode
        elif val < currentNode:
            return self._contains(val, currentNode.leftChild)
        else:
            return self._contains(val, currentNode.rightChild)


    # search for the minimum node

    def findMin(self):
        if self.root:
            return self._findMin(self.root)
        else:
            return None

    def _findMin(self, currentNode):
        if currentNode.hasLeftChild():
            return self._finMin(currentNode.leftChild)
        else:
            return currentNode


    # search for the maximum node

    def findMax(self):
        if self.root:
            return self._findMax(self.root)
        else:
            return None

    def _findMax(self, currentNode):
        if currentNode.hasRightChild():
            return _findMax(currentNode.rightChild)
        else:
            return currentNode

File: CPP/BinarySearchTree.cpp
#include <iostream>
using namespace std;

struct node {
    int data;
    node *left, *right;
};

node *newNode(int data){
	 node *n = new node;
	 n->data = data;
	 n->left = n->right = NULL;
	 return n;
}

void printInorder(node *root){
	 if(root == NULL)
	     return;
	 printInorder(root->left);
	 cout<<root->data<<" ";
	 printInorder(root->right);
}

void insert(node **root, int data){
     if(*root == NULL)
 		  *root = newNode(data);
    else{
		 if((*root)->data > data){
		     insert(&(*root)->left, data);		  	 
   		 }
   		 else
   		 	 insert(&(*root)->right, data);
    }
}
node *deleteNode(node *root, int key){
	if(root == NULL)
		return NULL;
	if(key < root->data){
		root->left = deleteNode(root->left, key);
	}
	else if(key > root->data){
		root->right = deleteNode(root->right, key);
	}
	else {
		if(root->left == NULL){
			node *res = root->right;
			delete(root);
			return res;
		}
		else if(root->right == NULL){
			node *res = root->left;
			delete(root);
			return res;
		}
		else {
			node *ptr = root->right;
			while(ptr->left){
				ptr = ptr->left;
			}
			root->data = ptr->data;
			root->right = deleteNode(root->right, ptr->data);
		}
	}
	return root;
}

int main(){
    /* Let us create following BST
              50
           /     \
          30      70
         /  \    /  \
       20   40  60   80 */
    struct node *root = NULL;
    insert(&root, 50);
    insert(&root, 30);
    insert(&root, 20);
    insert(&root, 40);
    insert(&root, 70);
    insert(&root, 60);
    insert(&root, 80);
  
    // print inoder traversal of the BST
    printInorder(root);
    getchar();
}
